src/charm.py

```diff
@@ -740,6 +730,153 @@ def _gen_wp_config(self):
+        This operation is idempotence.
+        """
+        if self._wordpress_service_exists():
+            self._container().stop("wordpress")
+
+    def _wp_is_installed(self):
+        """Check if WordPress is installed (check if WordPress related tables exist in database)"""
+        process = self._container().exec(["wp", "core", "is-installed"], timeout=60)
+        try:
+            process.wait()
+            return True
+        except ops.pebble.ExecError:
+            return False
+
+    def _wp_install_cmd(self):
+        """Generate wp-cli command used to install WordPress on database"""
+        initial_settings = yaml.safe_load(self.model.config["initial_settings"])
+        admin_user = initial_settings.get("user_name", "admin_username")
+        admin_email = initial_settings.get("admin_email", "name@example.com")
+        default_admin_password = self._replica_relation_data()["default_admin_password"]
+        admin_password = initial_settings.get("admin_password", default_admin_password)
+        return [
+            "wp",
+            "core",
+            "install",
+            "--url=localhost",
+            f"--title=The {self.model.config['blog_hostname'] or self.app.name} Blog",
+            f"--admin_user={admin_user}",
+            f"--admin_email={admin_email}",
+            f"--admin_password={admin_password}"
+        ]
+
+    def _wp_install(self):
+        """Install WordPress (create WordPress required tables in DB)"""
+        process = self._container().exec(self._wp_install_cmd(), timeout=60)
+        process.wait()
+
+    def _init_pebble_layer(self):
+        layer = {
+            "summary": "WordPress layer",
+            "description": "WordPress server",
+            "services": {
+                "wordpress": {
+                    "override": "replace",
+                    "summary": "WordPress server (apache)",
+                    "command": "apache2ctl -D FOREGROUND",
+                }
+            },
+        }
+        if not self._wordpress_service_exists():
+            self._container().add_layer("wordpress", layer, combine=True)
+
+    def _start_server(self):
+        """Start WordPress (apache) server. On leader unit, also make sure WordPress is installed
+
+        Check if the pebble layer has been added, then check the installation status of WordPress,
+        finally start the server. The installation process only run on the leader unit. This
+        operation is idempotence.
+        """
+        self._init_pebble_layer()
+        if self.unit.is_leader():
+            if not self._wp_is_installed():
+                self._wp_install()
+            if self._current_wp_config() is None:
+                # For security reasons, never start WordPress server if wp-config.php not exists
+                raise FileNotFoundError(
+                    "required file (wp-config.php) for starting WordPress server not exists"
+                )
+        self._container().start("wordpress")
```

Comment:
I'd go for self._container.pebble.replan_services() instead: https://github.com/canonical/operator/blob/main/ops/pebble.py#LC1625

Reply:
I think `replan_services` will only start the services if the services definition changes. Here the definition never changes after been set, but server will start and stop multiple times while running.